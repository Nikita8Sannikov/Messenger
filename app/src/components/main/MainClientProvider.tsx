import { createContext } from "react";

import App from "@components/main/app";
import MainClient from "@core/MainClient";

const mainClient = new MainClient();

// Инициализируем основной контекст всего приложения, куда прокидываем сущность MainClient
export const MainClientContext = createContext<MainClient>(mainClient);

// Основной компонент, главная задача которого ровно один раз инициализировать ядро бизнес-логики при открытии страницы пользователем
export default function MainClientProvider() {
	return <MainClientContext.Provider value={mainClient}>
		<App />
	</MainClientContext.Provider>;
};