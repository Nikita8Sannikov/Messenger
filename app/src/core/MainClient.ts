import EventEmitter from "eventemitter3";

import Logger from "@service/Logger";
import CatchErrors from "@core/CatchErrors";
import Request from "@core/Request";
import Socket from "@core/socket/Socket";
import ProfilesController from "@core/controllers/ProfilesController";
import { Profile } from "@core/models/Profile";
import MainApi from "@core/MainApi";
import { setAuth } from "@store/main/slice";
import { Pages } from "@custom-types/enums";
import { AppDispatch } from "@custom-types/redux.types";
import { MainClientEvents } from "@custom-types/events";
import { MY_ID } from "@utils/constants";

const logger = Logger.init();

// Класс, считающийся ядром бизнес логики на стороне клиента. Именно здесь происходит инициализация всех основных классов и вспомогательных сущностей
export default class MainClient extends EventEmitter {
    private readonly _request: Request;
    private readonly _catchErrors: CatchErrors;
    private readonly _profilesController: ProfilesController;
    private readonly _mainApi: MainApi;
    private readonly _socket: Socket;

    constructor(private readonly _dispatch: AppDispatch) {
        super();

        logger.debug("init");

        this._catchErrors = new CatchErrors(this._dispatch);
        this._request = new Request(this._catchErrors);

        this._profilesController = new ProfilesController(this._request, this._dispatch);
        
        this._socket = new Socket(this._dispatch);
        this._mainApi = new MainApi(this._request, this._dispatch);

        this._bindCatchErrorsListeners();
        this._bindProfileListeners();
        this._bindSocketListeners();
        this._bindMainApiListeners();
    }

    get mainApi() {
        return this._mainApi;
    }

    catchErrors(error: string) {
        this._catchErrors.catch(error);
    }

    getProfile(userId: string = MY_ID, showError: boolean = true) {
        return this._profilesController.getProfile(userId, showError) as Profile;
    }

    // Скачать файл с логами
    downloadLogFile() {
        logger.downloadToFile();
    }

    // Слушатели событый класса CatchErrors
    private _bindCatchErrorsListeners() {
        this._catchErrors.on(MainClientEvents.REDIRECT, (path: string) => {
            logger.debug(`MainClientEvents.REDIRECT: [path=${path}]`);
            this.emit(MainClientEvents.REDIRECT, path);
        });
    }

    // Слушатели событый класса ProfilesController
    private _bindProfileListeners() {
        this._profilesController.on(MainClientEvents.GET_ME, () => {
            logger.debug("MainClientEvents.GET_ME");
            this._getMe();
        });

        this._profilesController.on(MainClientEvents.ERROR, (error: string) => {
            logger.debug(`MainClientEvents.ERROR: [error=${error}]`);
            this.catchErrors(error);
        });
    }

    // Слушатели событый класса Socket
    private _bindSocketListeners() {
        this._socket.on(MainClientEvents.REDIRECT, (path: string) => {
            logger.debug(`MainClientEvents.REDIRECT: [path=${path}]`);
            this.emit(MainClientEvents.REDIRECT, path);
        });

        this._socket.on(MainClientEvents.ERROR, (error: string) => {
            logger.debug(`MainClientEvents.ERROR: [error=${error}]`);
            this.catchErrors(error);
        });
    }

    // Слушатели событый класса MainApi
    private _bindMainApiListeners() {
        this._mainApi.on(MainClientEvents.SIGN_IN, () => {
            logger.debug("MainClientEvents.SIGN_IN");

            // Необходимо обновить информацию о себе, так как после входа/регистрации информации о себе нет
            // Но при этом, уже созданы сущности "Профиль" и "Пользователь"
            // Если профиля не существует (только после выхода и без перезагрузки страницы, так как профиль удаляется только в этом случае), то необходимо заново его создать
            const myProfile = this.getProfile(MY_ID, false);

            myProfile
                ? myProfile.user.updateMe()
                : this._profilesController.addProfile();
        });

        this._mainApi.on(MainClientEvents.LOG_OUT, () => {
            logger.debug("MainClientEvents.LOG_OUT");

            this.emit(MainClientEvents.REDIRECT, Pages.signIn);
            this._dispatch(setAuth(false));
            this._socket.disconnect();
            this._profilesController.removeProfile();
        });
    }

    // Получили информацию о себе
    private _getMe() {
        this._dispatch(setAuth(true));
        this._initSocket();
        this._redirectHandler();
    }

    // Инициализация сокета
    private _initSocket() {
        logger.debug("_initSocket");

        const userInstance = this.getProfile().user;
        this._socket.init(userInstance.user);
    }

    // Редирект в зависимости от текущего урла
    private _redirectHandler() {
        switch (window.location.pathname) {
            case Pages.main:
            case Pages.signIn:
            case Pages.signUp:
                this.emit(MainClientEvents.REDIRECT, Pages.profile);
                break;
        }
    }
}