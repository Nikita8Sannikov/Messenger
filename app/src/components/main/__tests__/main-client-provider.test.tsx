import { useContext } from "react";
import { render } from "@testing-library/react";

import MainClientProvider, { MainClientContext } from "@components/main/MainClientProvider";
import MainClient from "@core/MainClient";

describe("MainClientProvider", () => {
	it("should render App component inside provider", () => {
		const { container } = render(<MainClientProvider />);
		expect(container.querySelector(".main-content")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it("should provide MainClient instance via context", () => {
		let contextValue: MainClient | undefined;
		function Consumer() {
			contextValue = useContext(MainClientContext);
			return null;
		}
		const mainClient = new MainClient();
		const { container } = render(
			<MainClientContext.Provider value={mainClient}>
				<Consumer />
			</MainClientContext.Provider>,
		);
		expect(contextValue).toBe(mainClient);
		expect(container).toMatchSnapshot();
	});
});
