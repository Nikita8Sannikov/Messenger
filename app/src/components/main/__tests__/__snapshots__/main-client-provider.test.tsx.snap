// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainClientProvider should provide MainClient instance via context 1`] = `<div />`;

exports[`MainClientProvider should render App component inside provider 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="root__wrapper root__no-auth"
    >
      <div
        class="root__wrapper__container"
      >
        <div
          class="main-content MuiBox-root css-0"
          data-testid="box-component"
        />
      </div>
    </div>
  </div>
</div>
`;
